### PR TITLE
API Update: Wrap PrefabSystemComponentInterface behavior

### PR DESCRIPTION
**Pull request by amzn-mike**

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp [hunk 1]
```diff
@@ -58,12 +60,20 @@ namespace AzToolsFramework
...
+            if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
             {
                 serialize->Class<PrefabSystemComponent, AZ::Component>()->Version(1);
             }
 
+            if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
+            {
+                behaviorContext->EBus<PrefabSystemComponentBus>("PrefabSystemComponentBus")
+                    ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Common)
+                    ->Attribute(AZ::Script::Attributes::Module, "types")
```
Comment (jackalbe) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp:72]: I think the Module should be all in the "entity" module if possible so that the PYI fall near each other

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp [hunk 2]
```diff
@@ -138,6 +148,26 @@ namespace AzToolsFramework
...
+            AZStd::vector<AZ::Entity*> entities;
+
+            for (const auto& entityId : entityIds)
+            {
+                AZ::Entity* entity = nullptr;
+                AZ::ComponentApplicationBus::BroadcastResult(entity, &AZ::ComponentApplicationBus::Events::FindEntity, entityId);
+
+                if (entity)
+                {
+                    entities.push_back(entity);
```
Comment (jackalbe) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp:163]: maybe at least an AZ_Warn() if the entity was not found? The scripter might have made a mistake in the code

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp [hunk 3]
```diff
@@ -138,6 +148,26 @@ namespace AzToolsFramework
...
+                }
+            }
+            
+            auto prefab = CreatePrefab(entities, {}, AZ::IO::PathView(AZStd::string_view(filePath)), nullptr, shouldCreateLinks);
```
Comment (jackalbe) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp:167]: I wonder if "entities is empty" is a significant state the caller would care about?

File: Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp [hunk 4]
```diff
@@ -21,6 +21,11 @@ namespace UnitTest
     {
     }
 
+    PrefabTestToolsApplication::PrefabTestToolsApplication(AZStd::string applicationName, int argc, char** argv)
```
Comment (jackalbe) [inline Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp:24]: will this affect other Prefab existing tests?

**Comment by jackalbe** [comment]
we need to include the Prefab coders on this too

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp (hunk 3 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp:167]: Hmm, that's hard to say.  It doesn't look like the CreatePrefab code prevents it.  Prefabs can be updated so maybe there's a use-case where they want to create an empty one and add entities later?

File: Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp (hunk 4 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp:24]: Appears not to, I believe I ran them all.  I can try running all of AzToolsFramework tests to make sure though

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp (hunk 3 above)
Comment (jackalbe) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp:170]: is there an API to "UpdatePreab" then?

Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp:170]: Good question.  There is, I think we need to add a task to expose it though.  The Instance has an AddEntity function, and PrefabDomUtils has StoreInstanceInPrefabDom and then PrefabSystemComponent has UpdatePrefabTemplate which takes a PrefabDom.

**Comment by AMZN-daimini** [comment]
The intent when we created the Prefab Public Handler was to have all public APIs go through that, and to not provide access to the system component directly.
Also, I believe Create Prefab was already exposed to Python via a bus that @chiyenteng is using to create automation test, so the same bus should be reused in that case.

File: Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp [hunk 5]
```diff
@@ -54,7 +59,17 @@ namespace UnitTest
...
     AZStd::unique_ptr<ToolsTestApplication> PrefabTestFixture::CreateTestApplication()
     {
-        return AZStd::make_unique<PrefabTestToolsApplication>("PrefabTestApplication");
+        constexpr size_t MaxCommandArgsCount = 8;
+        using FixedValueString = AZ::SettingsRegistryInterface::FixedValueString;
+        using ArgumentContainer = AZStd::fixed_vector<char*, MaxCommandArgsCount>;
+        // The first command line argument is assumed to be the executable name so add a blank entry for it
+        ArgumentContainer argContainer{ {} };
+
+        auto projectPathOverride = FixedValueString{ R"(--project-path=AutomatedTesting)" };
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp:68]: Is this hardcoding the project path to be 'AutomatedTesting' ? If so, what happens when a user creates a new project and decides to run existing prefab unit test?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h [hunk 6]
```diff
@@ -61,7 +62,23 @@ namespace AzToolsFramework
                 AZStd::unique_ptr<AZ::Entity> containerEntity = nullptr, bool ShouldCreateLinks = true) = 0;
         };
 
+        struct PrefabSystemComponentBehaviorContextInterface
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h:65]: Is this interface really necessary? From your other PR https://github.com/aws-lumberyard-dev/o3de/pull/208/files , it looks like the virtual method can directly be added to the Bus class? 
This interface would be useful if we want users to use this instead of ebuses. But we already have PrefabSystemComponentInterface. So, this seems like extra baggage that can be cut off?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.h [hunk 7]
```diff
@@ -202,6 +203,8 @@ namespace AzToolsFramework
                 AZ::IO::PathView filePath, AZStd::unique_ptr<AZ::Entity> containerEntity = nullptr,
                 bool ShouldCreateLinks = true) override;
 
+            TemplateId CreatePrefabTemplate(const AZStd::vector<AZ::EntityId>& entities, const AZStd::string& filePath, bool shouldCreateLinks) override;
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.h:206]: I suggest making this a private method to prevent callers from directly instantiating this class and calling this method. This should only be callable through the ebus.
Also consider adding a comment section to indicate that this is a bus implementation. It can be expanded to any other methods you add to this bus.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp (hunk 3 above)
Comment (srikappa-amzn) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp:170]: Entities being empty is a valid use case. Because a prefab can be created with 0 entities and 'n' number of nested prefab instances. If both entities and nested instances are empty, then you will be creating an empty prefab.
Although one can even argue that they just want to create a placeholder prefab and add any entities/nestedInstances later.

File: Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp (hunk 5 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/Tests/Prefab/PrefabTestFixture.cpp:68]: Seems it's not needed anymore.  There's a bunch of tests that complain about not finding the project path but its just a warning that doesn't fail the test so I'll just leave this out.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h (hunk 6 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h:65]: Removed

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.h (hunk 7 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.h:206]: Fixed

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h [hunk 8]
```diff
@@ -61,7 +62,18 @@ namespace AzToolsFramework
                 AZStd::unique_ptr<AZ::Entity> containerEntity = nullptr, bool ShouldCreateLinks = true) = 0;
         };
 
-
+        struct PrefabSystemComponentEbusTraits : AZ::EBusTraits
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h:65]: Having this bus in this class would have been better if this is simply a bus wrapping the PrefabSystemComponentInterface. [Here's an example](https://github.com/o3de/o3de/blob/5e04c3737ffdec3e3b710968b8d7b835c91885e1/Code/Framework/AzCore/AzCore/Component/ComponentApplicationBus.h#L205-L221)

I suggest moving this ebus and its implementation into separate files for the below reasons:
1.If we ever needed to provide an ebus that wraps the PrefabSystemComponentInterface, this bus would conflict with it. PrefabSystemComponentBus should behave just like PrefabSystemComponentInterface except for it being a bus vs interface
2. PrefabSystemComponent is already pretty big enough. By moving the bus and implementation to separate files, we keep the code manageable
3. The new bus should be called something like 'ProceduralPrefabRequestBus' and comments should be added on who should be calling this bus. This clearly separates the roles of PrefabSystemComponent and ProceduralPrefabRequestBusHandler, which utilizes PrefabSystemComponent to satisfy a unique set of requirements.

An example of this pattern is the PrefabPublicRequestHandler, which doesn't wrap the PrefabPublicInterface but instead utilizes it to satisfy the requirements of automated tests.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h [hunk 9]
```diff
@@ -61,7 +62,18 @@ namespace AzToolsFramework
...
+        {
+            using MutexType = AZ::NullMutex;
+            static const AZ::EBusHandlerPolicy HandlerPolicy = AZ::EBusHandlerPolicy::Single;
+            static const AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::Single;
+            
+            virtual TemplateId CreatePrefabTemplate(
+                const AZStd::vector<AZ::EntityId>& entityIds, const AZStd::string& filePath, bool shouldCreateLinks) = 0;
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h:72]: 'shouldCreateLinks' is used only when nested prefabs are used. In your case, you are simply providing a list of entities to be added to the prefab. So this flag need not be provided as a function parameter. You can instead hardcode this to (true/false -- doesn't really matter since it won't be used) in the call to PrefabSystemComponent::CreatePrefab()

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingHandler.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        public:
+            static void Reflect(AZ::ReflectContext* context);
+
+            PrefabSystemScriptingHandler() = default;
+
+            void Connect(PrefabSystemComponentInterface* prefabSystemComponentInterface);
+            void Disconnect();
+
+        private:
+            AZ_DISABLE_COPY(PrefabSystemScriptingHandler);
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingHandler.h:28]: You can use AZ_DISABLE_COPY_MOVE here to disable the move as well. I doubt that we will need it for this class

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h [hunk 11]
```diff
@@ -14,6 +14,7 @@
 #include <AzToolsFramework/Prefab/PrefabIdTypes.h>
 #include <AzToolsFramework/Prefab/Template/Template.h>
 #include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <AzCore/EBus/EBus.h>
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h:17]: can be deleted now I guess?

**Review by srikappa-amzn — APPROVED** [review]
Looks good to me. Left some minor comments.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingBus.h [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    namespace Prefab
+    {
+        // Bus that exposes a script-friendly interface to the PrefabSystemComponent
+        struct PrefabSystemScriptingEbusTraits : AZ::EBusTraits
+        {
+            using MutexType = AZ::NullMutex;
+            static const AZ::EBusHandlerPolicy HandlerPolicy = AZ::EBusHandlerPolicy::Single;
+            static const AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::Single;
+
+            virtual TemplateId CreatePrefabTemplate(
```
[the remark this answers is not in the document]
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingBus.h:30]: I'm not sure what I can put here.   CreatePrefab doesn't have any comments on it for me to copy.  //Creates a prefab template isn't considered useful commenting so I left it out.  Thoughts?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingHandler.cpp [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Common)
+                ->Attribute(AZ::Script::Attributes::Module, "prefab")
+                ->Attribute(AZ::Script::Attributes::Category, "Prefab")
+                ->Event("CreatePrefab", &PrefabSystemScriptingBus::Events::CreatePrefabTemplate);
+        }
+    }
+
+    void PrefabSystemScriptingHandler::Connect(PrefabSystemComponentInterface* prefabSystemComponentInterface)
+    {
+        AZ_Assert(prefabSystemComponentInterface != nullptr, "prefabSystemComponentInterface must not be null");
```
[the remark this answers is not in the document]
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingHandler.cpp:31]: Well I did that on purpose to point out the variable which is null...

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingHandler.h (hunk 10 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemScriptingHandler.h:28]: Probably doesn't matter one way or another in the end, as its highly unlikely this will ever be copied/moved, but there's not really an issue with it being moved.  It's a singleton so copying it is actually a problem.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h (hunk 11 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponentInterface.h:17]: Removed

File: Code/Framework/AzToolsFramework/Tests/Prefab/PrefabScriptingTests.cpp [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+        behaviorContext->Property("g_globalTemplateId", BehaviorValueProperty(&g_globalTemplateId));
+
+        g_globalTemplateId = TemplateId{};
+        
+        sc.BindTo(behaviorContext);
+        sc.Execute(R"LUA(
+            my_id = EditorEntityUtilityBus.Broadcast.CreateEditorReadyEntity("test")
+            entities = vector_EntityId(my_id)
+            g_globalTemplateId = PrefabSystemScriptingBus.Broadcast.CreatePrefab(entities, "test.prefab", true)
```
Comment (AMZN-stankowi) [inline Code/Framework/AzToolsFramework/Tests/Prefab/PrefabScriptingTests.cpp:34]: It would be useful to have a few more tests on invalid/edge cases. Empty entity list, null entity list, invalid or empty prefab name string.

**Review by AMZN-stankowi — APPROVED** [review]

File: Code/Framework/AzToolsFramework/Tests/Prefab/PrefabScriptingTests.cpp (hunk 14 above)
Comment (amzn-mike) [reply Code/Framework/AzToolsFramework/Tests/Prefab/PrefabScriptingTests.cpp:34]: Lua won't let you pass in nil here but I can do the other tests.  I think Im going to do this as a separate PR though because so far this is requiring merging in 2 other branches.
